sdks/python/apache_beam/runners/portability/fn_api_runner_test.py

```diff
@@ -1083,6 +1131,34 @@ def __reduce__(self):
 def _unpickle_element_counter(name):
   return _pickled_element_counters[name]
 
+class EventRecorder(object):
+  """Used to be registered as a callback in bundle finalization.
+
+  The reason why records are written into a tmp file is, the in-memory dataset
+  cannot keep callback records when passing into one DoFn.
+  """
+  def __init__(self, tmp_dir):
+    self.tmp_dir = tmp_dir
+    self.file_path = os.path.join(self.tmp_dir, uuid.uuid4().hex + '.txt')
+
+  def record(self, content):
+    tmp_file = open(self.file_path, 'a')
```

Comment:
Nit: preferable to use with statements for file. E.g.

with open(file_path, 'w') as f:
    f.write(...)